[Core][Model] Warning of infinte recursion

I get the following warning:

~~~
/home/vicente/src/Kratos/kratos/containers/model.cpp:174:1: warning: all paths through this function will call itself [-Winfinite-recursion]
~~~

--- a/kratos/containers/model.cpp
+++ b/kratos/containers/model.cpp
@@ -172,7 +172,78 @@ ModelPart& Model::GetModelPart(const std::string& rFullModelPartName)
 
 const ModelPart& Model::GetModelPart(const std::string& rFullModelPartName) const
 {
-    return this->GetModelPart(rFullModelPartName);
+    KRATOS_TRY
+
+    KRATOS_ERROR_IF( rFullModelPartName.empty() ) << "Attempting to find a "
+        << "ModelPart with empty name (\"\")!" << std::endl;
+
+    std::vector< std::string > subparts_list = SplitSubModelPartHierarchy(rFullModelPartName);
+
+
+    if(subparts_list.size() == 1) //it is a root model part
+    {
+        auto search = mRootModelPartMap.find(subparts_list[0]);
+        if(search != mRootModelPartMap.end())
+        {
+            return *(search->second);
+        }
+        else //let's also search it as a flat name - a feature that SHOULD BE DEPRECATED
+        {
+
+            for(auto it = mRootModelPartMap.begin(); it!=mRootModelPartMap.end(); it++)
+            {
+                ModelPart* pmodel_part = RecursiveSearchByName(subparts_list[0], (it->second.get()));
+                if (pmodel_part != nullptr) { //give back the first one that was found
+                    // Get the names of the parent-modelparts to print them in the warning
+                    std::vector<std::string> model_part_names;
+                    GetNameWithAscendants(*pmodel_part, model_part_names);
+
+                    std::stringstream msg;
+                    msg << model_part_names[0];
+                    for (std::size_t i=1; i<model_part_names.size(); ++i) {
+                        msg << "." << model_part_names[1];
+                    }
+
+                    KRATOS_INFO("Model") << "DEPREATION_WARNING: The ModelPart \"" // TODO make a warning at some point
+                        << subparts_list[0] << "\"\nis retrieved from the Model by using the "
+                        << "flat-map!\nPlease prepend the Parent-ModelPart-Names like this:\n\""
+                        << msg.str() << "\"" << std::endl;
+
+                    return *pmodel_part;
+                }
+            }
+
+            //if we are here we did not find it
+            KRATOS_ERROR << "The ModelPart named : \"" << subparts_list[0]
+                    << "\" was not found either as root-ModelPart or as a flat name. The total input string was \""
+                    << rFullModelPartName << "\"" << std::endl;
+        }
+    }
+    else //it is a submodelpart with the full name provided
+    {
+        auto search = mRootModelPartMap.find(subparts_list[0]);
+        if(search != mRootModelPartMap.end())
+        {
+            ModelPart* p_model_part = (search->second).get();
+            for(unsigned int i=1; i<subparts_list.size(); ++i)
+            {
+                KRATOS_ERROR_IF_NOT(p_model_part->HasSubModelPart(subparts_list[i]))
+                    << "The ModelPart named : \"" << subparts_list[i]
+                    << "\" was not found as SubModelPart of : \""
+                    << subparts_list[i-1] << "\". The total input string was \""
+                    << rFullModelPartName << "\"" << std::endl;
+                p_model_part = &p_model_part->GetSubModelPart(subparts_list[i]);
+            }
+            return *p_model_part;
+        }
+        else
+        {
+            KRATOS_ERROR << "root model part " << rFullModelPartName << " not found" << std::endl;
+        }
+
+    }
+
+    KRATOS_CATCH("")
 }
 
 bool Model::HasModelPart(const std::string& rFullModelPartName) const
@@ -274,9 +345,8 @@ std::vector<std::string> Model::SplitSubModelPartHierarchy(const std::string& rF
     return rSubPartsList;
 }
 
-ModelPart* Model::RecursiveSearchByName(const std::string& ModelPartName, ModelPart* pModelPart)
+ModelPart* Model::RecursiveSearchByName(const std::string& ModelPartName, ModelPart* pModelPart) const
 {
-
     for(auto& part : pModelPart->SubModelParts())
     {
         if(part.Name() == ModelPartName)

--- a/kratos/containers/model.h
+++ b/kratos/containers/model.h
@@ -248,7 +248,7 @@ private:
      * @param pModelPart Pointer of the model part where search recursively
      * @return The pointer of the model part of interest
      */
-    ModelPart* RecursiveSearchByName(const std::string& rModelPartName, ModelPart* pModelPart);
+    ModelPart* RecursiveSearchByName(const std::string& rModelPartName, ModelPart* pModelPart) const;
 
     /**
      * @brief This method splits the name of the model part using "." to define the hierarchy

